pkg/scheduler/actions/reserve/reserve.go

```diff
@@ -0,0 +1,249 @@
+	queues := util.NewPriorityQueue(ssn.QueueOrderFn)
+	jobsMap := map[api.QueueID]*util.PriorityQueue{}
+
+	for _, job := range ssn.Jobs {
+		if queue, found := ssn.Queues[job.Queue]; found {
+			queues.Push(queue)
+		} else {
+			glog.V(4).Infof("Skip adding job <%s/%s> because its queue %s is not found",
+				job.Namespace, job.Name, job.Queue)
+			continue
+		}
+		if !isJobStarving(job, threshold) {
+			glog.V(4).Infof("Skip adding job <%s/%s> because it is not starving",
+				job.Namespace, job.Name, job.Queue)
+			continue
+		}
+
+		if _, found := jobsMap[job.Queue]; !found {
+			jobsMap[job.Queue] = util.NewPriorityQueue(ssn.JobOrderFn)
+		}
+
+		glog.V(4).Infof("Added Job <%s/%s> into Queue <%s>", job.Namespace, job.Name, job.Queue)
+		jobsMap[job.Queue].Push(job)
+	}
+
+	pendingTasks := map[api.JobID]*util.PriorityQueue{}
+	allNodes := util.GetNodeList(ssn.Nodes)
+	predicateFn := func(task *api.TaskInfo, node *api.NodeInfo) error {
+		// a node is elligible for starving job if its allocatable >= task.InitResreq
+		if !task.InitResreq.LessEqual(node.Allocatable) {
+			return fmt.Errorf("task <%s/%s> ResourceFit failed on node <%s>",
+				task.Namespace, task.Name, node.Name)
+		}
+		return ssn.PredicateFn(task, node)
+	}
+
+	// (over-)allocate resource to starving jobs
+	for {
+		if queues.Empty() {
+			break
+		}
+
+		queue := queues.Pop().(*api.QueueInfo)
+		if ssn.Overused(queue) {
+			glog.V(3).Infof("Queue <%s> is overused, ignore it.", queue.Name)
+			continue
+		}
+
+		jobs, found := jobsMap[queue.UID]
+		if !found || jobs.Empty() {
+			glog.V(4).Infof("Cannot find jobs for queue %s.", queue.Name)
+			continue
+		}
+
+		// collect all pending tasks of the starving job
+		job := jobs.Pop().(*api.JobInfo)
+		if _, found := pendingTasks[job.UID]; !found {
+			tasks := util.NewPriorityQueue(ssn.TaskOrderFn)
+			for _, task := range job.TaskStatusIndex[api.Pending] {
+				// Skip BestEffort task in 'allocate' action.
+				if task.Resreq.IsEmpty() {
+					glog.V(4).Infof("Task <%v/%v> is BestEffort task, skip it.",
+						task.Namespace, task.Name)
+					continue
+				}
+
+				tasks.Push(task)
+			}
+			pendingTasks[job.UID] = tasks
+		}
+		tasks := pendingTasks[job.UID]
+
+		glog.V(3).Infof("Trying to reserve resource for %d tasks of starving Job <%v/%v>",
+			tasks.Len(), job.Namespace, job.Name)
+
+		for !tasks.Empty() {
+			task := tasks.Pop().(*api.TaskInfo)
+
+			glog.V(3).Infof("There are <%d> nodes for Job <%v/%v>",
+				len(ssn.Nodes), job.Namespace, job.Name)
+
+			if len(job.NodesFitDelta) > 0 {
+				job.NodesFitDelta = make(api.NodeResourceMap)
+			}
+
+			// Going through all the nodes.
+			// When allocating for starving, node needs
+			// to be offered in consistent order
+			// If no node is available, task will not be allocated
+			// in any case
+			nodes := util.PredicateNodes(task, allNodes, predicateFn)
+			if len(nodes) == 0 {
+				break
+			}
+
+			taskAllocated := false
+			for _, node := range nodes {
```

Comment:
We need to order nodes in our own way to maintain the stable node ordering between scheduling rounds.